src/Microsoft.Health.Dicom.Functions/Update/UpdateDurableFunction.Activity.cs

```diff
@@ -125,6 +126,95 @@ public async Task<IEnumerable<InstanceMetadata>> UpdateInstanceWatermarkV2Async(
         return updatedInstances;
     }
 
+    /// <summary>
+    /// Asynchronously batches the instance watermarks and calls the update instance.
+    /// </summary>
+    /// <param name="arguments">BatchUpdateArguments</param>
+    /// <param name="logger">A diagnostic logger.</param>
+    /// <returns>
+    /// The result of the task contains the updated instances with file properties representing newly created blobs and any error.
+    /// </returns>
+    /// <exception cref="ArgumentNullException">
+    /// <paramref name="arguments"/> or <paramref name="logger"/> is <see langword="null"/>.
+    /// </exception>
+    [FunctionName(nameof(UpdateInstanceBlobsV3Async))]
+    public async Task<UpdateInstanceResponse> UpdateInstanceBlobsV3Async(
+        [ActivityTrigger] UpdateInstanceBlobArgumentsV2 arguments,
+        ILogger logger)
+    {
+        EnsureArg.IsNotNull(arguments, nameof(arguments));
+        EnsureArg.IsNotNull(arguments.ChangeDataset, nameof(arguments.ChangeDataset));
+        EnsureArg.IsNotNull(arguments.InstanceMetadataList, nameof(arguments.InstanceMetadataList));
+        EnsureArg.IsNotNull(arguments.Partition, nameof(arguments.Partition));
+        EnsureArg.IsNotNull(logger, nameof(logger));
+
+        DicomDataset datasetToUpdate = GetDeserialzedDataset(arguments.ChangeDataset);
+
+        int processed = 0;
+
+        logger.LogInformation("Beginning to update all instance blobs, Total count {TotalCount}", arguments.InstanceMetadataList.Count);
+
+        ConcurrentBag<InstanceMetadata> updatedInstances = new ConcurrentBag<InstanceMetadata>();
+        var errors = new ConcurrentBag<string>();
+
+        while (processed < arguments.InstanceMetadataList.Count)
```

Comment:
If all of the processing is happening within `Parallel.ForEachAsync`, then you don't need to further batch the data. You can limit the concurrency by setting `MaxDegreeOfParallelism` instead. So simply pass the entire list of metadata to `Parallel.ForEachAsync`

Reply:
It was the existing code, missed refactoring. Corrected